### PR TITLE
Transfer admin permissions to another member of institution

### PR DESCRIPTION
**Pull request by Luiz-FS**
**Feature/Bug description:** Create on the server the possibility of an administrator transferring his position to another member of the same institution.

**Solution:** I created on the server a model inviteUserAdm, responsible for creating the invitation to the administration transfer, it verifies if the user for whom it is being asked to be the new administrator is a member of the institution. I also created the InviteUserAdmHandler handler, which managed the invitation when the user accepts or rejects, transferring the permissions, and, to transfer all the permissions from one administrator to another, I created the TransferAdminPermissionsHandler handler, which takes all permissions regarding the institution and its hierarchy transfer to the new administrator.

**TODO/FIXME:** n/a

### CODE REVIEW

File: backend/handlers/invite_user_adm_handler.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            "This invitation has already been accepted", 
+            NotAuthorizedException)
+        
+        Utils._assert(
+            invite.status == 'rejected', 
+            "This invitation has already been rejected", 
+            NotAuthorizedException)
+        
+        Utils._assert(
+            invite.make()['type_of_invite'] != 'INVITE_USER_ADM',
```
Comment (andreldsa) [inline backend/handlers/invite_user_adm_handler.py]: Why you had to call make()?

File: backend/handlers/invite_user_adm_handler.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        Utils._assert(
+            invite.make()['type_of_invite'] != 'INVITE_USER_ADM', 
+            "invitation type not allowed", 
+            NotAuthorizedException)
+
+        invite.change_status('accepted')
+        actual_admin = invite.admin_key.get()
+        institution = invite.institution_key.get()
+
+        user.institutions_admin.append(institution.key)
```
Comment (andreldsa) [inline backend/handlers/invite_user_adm_handler.py]: You can't change model directly on the handler, you should call a model function.

File: backend/handlers/invite_user_adm_handler.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        actual_admin.institutions_admin.remove(institution.key)
```
Comment (andreldsa) [inline backend/handlers/invite_user_adm_handler.py]: same

File: backend/handlers/invite_user_adm_handler.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+        institution.put()
```
Comment (andreldsa) [inline backend/handlers/invite_user_adm_handler.py]: Those three lines must be in a transaction.

File: backend/test/model_test/invite_user_adm_test.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        cls.test.init_memcache_stub()
+    
+    def test_create(self):
+        """Test create new invite."""
+        institution = mocks.create_institution()
+        admin = mocks.create_user()
+        new_admin = mocks.create_user()
+
+        institution.admin = admin.key
+        institution.members = [admin.key, new_admin.key]
```
Comment (andreldsa) [inline backend/test/model_test/invite_user_adm_test.py]: Never change the model outside the model, we have functions to do that, like add_member

File: backend/test/model_test/invite_user_adm_test.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        admin.institutions_admin = [institution.key]
```
Comment (andreldsa) [inline backend/test/model_test/invite_user_adm_test.py]: sab=me

File: backend/test/model_test/invite_user_adm_test.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+        cls.test.init_datastore_v3_stub(consistency_policy=cls.policy)
+        cls.test.init_memcache_stub()
+    
+    def test_create(self):
+        """Test create new invite."""
+        institution = mocks.create_institution()
+        admin = mocks.create_user()
+        new_admin = mocks.create_user()
+
+        institution.admin = admin.key
```
Comment (andreldsa) [inline backend/test/model_test/invite_user_adm_test.py]: same

File: backend/test/model_test/invite_user_adm_test.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        )
+    
+    def test_create_invite_with_invitee_not_a_member(self):
+        """Test create invite whith invitee not a member."""
+        institution = mocks.create_institution()
+        admin = mocks.create_user()
+        new_admin = mocks.create_user()
+
+        institution.admin = admin.key
+        institution.members = [admin.key]
```
Comment (andreldsa) [inline backend/test/model_test/invite_user_adm_test.py]: same ...

File: backend/utils.py [hunk 9]
```diff
@@ -260,3 +261,39 @@ def to_int(value, exception, message_exception):
         raise exception(message_exception)
 
     return value
+
+
+def makeUser(user, request):
```
Comment (andreldsa) [inline backend/utils.py]: Why utils? Why not user.py?

File: backend/utils.py [hunk 10]
```diff
@@ -260,3 +261,39 @@ def to_int(value, exception, message_exception):
...
+            Utils.toJson(institution.get())
+        )
+    user_json['follows'] = [institution_key.get().make(
+        ['acronym', 'photo_url', 'key', 'parent_institution']) for institution_key in user.follows]
+    user_json['institution_profiles'] = [profile.make() 
+        for profile in user.institution_profiles]
+    return user_json
+    
+
+def get_all_hierarchy_admin_permissions(institution_key, permissions_dict=None):
```
Comment (andreldsa) [inline backend/utils.py]: permissions*, remove de _dic

Comment (andreldsa) [inline backend/utils.py]: Create a well explained doc for this function.

Comment (andreldsa) [inline backend/utils.py]: This function does not feat the context of utils, should be in another file.

File: backend/worker.py [hunk 11]
```diff
@@ -238,6 +239,42 @@ def post(self):
             invite = ndb.Key(urlsafe=key).get()
             invite.send_invite(host, current_institution)
 
+
+class TransferAdminPermissionsHandler(BaseHandler):
+
+    def add_permission(self, user, permissions):
```
Comment (andreldsa) [inline backend/worker.py]: doc

File: backend/worker.py [hunk 12]
```diff
@@ -238,6 +239,42 @@ def post(self):
...
+        for permission in permissions:
+            if permission in user.permissions:
+                user.permissions[permission].update(permissions[permission])
+            else:
+                user.permissions.update({permission: permissions[permission]})
+    
+    def remove_permissions(self, user, permissions, institution_key):
```
Comment (andreldsa) [inline backend/worker.py]: doc and refact, too much complex.

File: backend/worker.py [hunk 13]
```diff
@@ -238,6 +239,42 @@ def post(self):
...
+        for permission in permissions:
+            if permission in user.permissions:
+                for institution in permissions[permission]:
+                    if institution != institution_key and ndb.Key(urlsafe=institution) not in user.institutions_admin:
+                        del user.permissions[permission][institution]
```
Comment (andreldsa) [inline backend/worker.py]: You shouldn't modify the user inside a handler, we have functions to do that.

File: backend/worker.py [hunk 14]
```diff
@@ -238,6 +239,42 @@ def post(self):
...
+                    elif institution == institution_key:
+                        del user.permissions[permission][institution]
```
Comment (andreldsa) [inline backend/worker.py]: same

File: backend/worker.py [hunk 15]
```diff
@@ -238,6 +239,42 @@ def post(self):
...
+        permissions = get_all_hierarchy_admin_permissions(institution_key)
+        institution = ndb.Key(urlsafe=institution_key).get()
+        admin = institution.admin.get()
+        new_admin = ndb.Key(urlsafe=user_key).get()
+        institution.admin = new_admin.key
+
+        self.add_permission(new_admin, permissions)
+        self.remove_permissions(admin, permissions, institution_key)
+
+        new_admin.put()
```
Comment (andreldsa) [inline backend/worker.py]: Those three lines must be in a transaction or it can cause a inconsistency.

**Comment by andreldsa** [comment]
Great work, with some tweaks that will be really good.

File: backend/test/invite_user_adm_hander_test.py [hunk 16]
```diff
@@ -0,0 +1,460 @@
...
+        self.assertEquals(
+            invite.status, 
+            'accepted',
+            'Invitation status must be equal to accepted!'
+        )
+    
+    @patch.object(Invite, 'send_notification')
+    @patch('handlers.invite_user_adm_handler.enqueue_task')
+    @patch('utils.verify_token', return_value={'email': 'usr_test@test.com'})
+    def test_put_invite_in_hierarchy(self, verify_token, enqueue_task, send_notification):
```
Comment (mayzabeel) [inline backend/test/invite_user_adm_hander_test.py]: :1st_place_medal:  que teste!

File: backend/test/model_test/invite_user_adm_test.py (hunk 6 above)
Comment (Luiz-FS) [reply backend/test/model_test/invite_user_adm_test.py]: What?

File: backend/handlers/invite_user_adm_handler.py [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+class InviteUserAdmHandler(BaseHandler):
+    """Invite User Admin Handler."""
+
+    @json_response
+    @login_required
+    def put(self, user, invite_key):
+        """Handler of accept invite."""
+        invite = ndb.Key(urlsafe=invite_key).get()
+
+        Utils._assert(
```
[the remark this answers is not in the document]
Comment (Luiz-FS) [reply backend/handlers/invite_user_adm_handler.py]: Done

File: backend/test/model_test/invite_user_adm_test.py (hunk 5 above)
Comment (Luiz-FS) [reply backend/test/model_test/invite_user_adm_test.py]: Done

File: backend/models/user.py [hunk 18]
```diff
@@ -211,6 +211,12 @@ def add_institution(self, institution_key):
             self.add_permission("publish_survey", institution_key.urlsafe())
             self.put()
 
+    def add_institution_admin(self, institution_key):
+        """Add a institution admin to user."""
```
Comment (andreldsa) [inline backend/models/user.py]: What? Explain better. It seems that add a User to that list but receives a institituion_key.

File: backend/worker.py [hunk 19]
```diff
@@ -10,12 +10,45 @@
...
+    returns a dictionary of filtered permissions that must be removed 
+    from the user according to the rules applied in the 'should_remove' method.
+
+    Arguments:
+    user -- User to verify if permission must be removed
+    permissions -- Permissions to filter
+    institution_key -- Key of the institution from which the transfer of permissions is being made.
+    """
+    permissions_filtered = {}
+    for permission, insts in permissions.items():
```
Comment (andreldsa) [inline backend/worker.py]: insts -> institutions*

File: backend/worker.py [hunk 20]
```diff
@@ -10,12 +10,45 @@
...
+            inst_keys = [inst for inst in insts if should_remove(user, inst, institution_key)]
```
Comment (andreldsa) [inline backend/worker.py]: inst -> institution

Use the real name, avoid contractions and misunderstandings.

File: backend/worker.py [hunk 21]
```diff
@@ -10,12 +10,45 @@
...
+            permissions_filtered[permission] = inst_keys
```
Comment (andreldsa) [inline backend/worker.py]: institution_keys

**Pull request merged**